### PR TITLE
[Fix #1881] Add `cider-cljs-*-repl` defcustom and hook `boot-cljs-repl`

### LINKED ISSUE #1881: Error starting CLJS REPL with Boot project
## Expected behavior

I should get a working CLJS REPL when I run `cider-jack-in-clojurescript`.

## Actual behavior

I get this stack trace:

```

  Show: Clojure Java REPL Tooling Duplicates All  (13 frames hidden)

1. Unhandled java.lang.ClassNotFoundException
   cemerick.piggieback

       URLClassLoader.java:  381  java.net.URLClassLoader/findClass
   DynamicClassLoader.java:   69  clojure.lang.DynamicClassLoader/findClass
          ClassLoader.java:  424  java.lang.ClassLoader/loadClass
   DynamicClassLoader.java:   77  clojure.lang.DynamicClassLoader/loadClass
          ClassLoader.java:  357  java.lang.ClassLoader/loadClass
                Class.java:   -2  java.lang.Class/forName0
                Class.java:  348  java.lang.Class/forName
                   RT.java: 2154  clojure.lang.RT/classForName
                   RT.java: 2163  clojure.lang.RT/classForName
             Compiler.java: 1017  clojure.lang.Compiler$HostExpr/maybeClass
             Compiler.java:  795  clojure.lang.Compiler$HostExpr/access$600
             Compiler.java: 6662  clojure.lang.Compiler/macroexpand1
             Compiler.java: 6692  clojure.lang.Compiler/macroexpand
             Compiler.java: 6766  clojure.lang.Compiler/eval
             Compiler.java: 6745  clojure.lang.Compiler/eval
                  core.clj: 3081  clojure.core/eval
                  main.clj:  240  clojure.main/repl/read-eval-print/fn
                  main.clj:  240  clojure.main/repl/read-eval-print
                  main.clj:  258  clojure.main/repl/fn
                  main.clj:  258  clojure.main/repl
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   87  clojure.tools.nrepl.middleware.interruptible-eval/evaluate/fn
                  AFn.java:  152  clojure.lang.AFn/applyToHelper
                  AFn.java:  144  clojure.lang.AFn/applyTo
                  core.clj:  630  clojure.core/apply
                  core.clj: 1868  clojure.core/with-bindings*
               RestFn.java:  425  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   85  clojure.tools.nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  222  clojure.tools.nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
    interruptible_eval.clj:  190  clojure.tools.nrepl.middleware.interruptible-eval/run-next/fn
                  AFn.java:   22  clojure.lang.AFn/run
   ThreadPoolExecutor.java: 1142  java.util.concurrent.ThreadPoolExecutor/runWorker
   ThreadPoolExecutor.java:  617  java.util.concurrent.ThreadPoolExecutor$Worker/run
               Thread.java:  745  java.lang.Thread/run
```

The REPL itself seems to sort of work, but any CIDER commands fail with a fresh exception.

## Steps to reproduce the problem

1. Run `boot -d seancorfield/boot-new new -t hoplon -n address-book`
2. Open the one CLJS file in the template project:`C-x C-f address-book/cljs/index.cljs.hl`
3. Run `M-x clojurescript-mode` (if it isn't set up to `.cljs.hl` files already).
4. Press `C-c M-J`

### CIDER version information

```
;; CIDER 0.14.0 (Berlin), nREPL 0.2.12
;; Clojure 1.8.0, Java 1.8.0_51
```

### Lein/Boot version

Boot 2.6.0

### Emacs version

Emacs 25.5.1

### Operating system

macOS 10.12.1 "Sierra"

### PR DESCRIPTION
**Pull request by arichiardi**
This patch firstly adds two more defcustoms, `cider-cljs-boot-repl` and
`cider-cljs-gradle-repl`, that will be queried by the
`cider-cljs-repl-form` function on `cider-jack-in-clojurescript`

Secondly, it adds a clause in `cider--cljs-repl-types` that allows to
automatically call `adzerk.boot-cljs-repl/start-repl` in case `boot` is
the project type.

**Replace this placeholder text with a summary of the changes in your PR.
The more detailed you are, the better.**

-----------------

Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [X] The commits are consistent with our [contribution guidelines][1]
- [X] All tests are passing (`make test`)
- [X] The new code is not generating bytecode or `M-x checkdoc` warnings
- [X] You've updated the changelog (if adding/changing user-visible functionality)
- [X] You've updated the readme (if adding/changing user-visible functionality)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -29,6 +29,7 @@
 * [#1875](https://github.com/clojure-emacs/cider/issues/1875): Ensure cljc buffers can load buffer into both clj and cljs repls.
 * [#1897](https://github.com/clojure-emacs/cider/issues/1897): Bind TAB in stacktrace buffers in the terminal.
 * [#1895](https://github.com/clojure-emacs/cider/issues/1895): Connect to the same host:port after `cider-restart` if the connection was established with `cider-connect`.
+* [#1881](https://github.com/clojure-emacs/cider/issues/1881): Error starting ClojureScript Repl with Boot project
```
Comment (bbatsov) [inline CHANGELOG.md]: You should actually describe the solution of the problem. :-) And add a final `.`.

File: cider-mode.el [hunk 2]
```diff
@@ -186,9 +186,11 @@ Returns to the buffer in which the command was invoked."
     ("Clojurescript"
      ["Start a Clojure REPL, and a ClojureScript REPL" cider-jack-in-clojurescript
       :help "Starts an nREPL server, connects a Clojure REPL to it, and then a ClojureScript REPL.
-Configure `cider-cljs-lein-repl' to change the ClojureScript REPL to use."]
+Configure `cider-cljs-*-repl' to change the ClojureScript REPL to use for your build tool."]
      ["Create a ClojureScript REPL from a Clojure REPL" cider-create-sibling-cljs-repl]
-     ["Configure the ClojureScript REPL to use" (customize-variable 'cider-cljs-lein-repl)])
+     ["Form for launching a lein ClojureScript REPL" (customize-variable 'cider-cljs-lein-repl)]
```
Comment (bbatsov) [inline cider-mode.el]: for launching a ClojureScript REPL via Leiningen (and so on)

File: cider-mode.el [hunk 3]
```diff
@@ -342,7 +344,7 @@ Configure `cider-cljs-lein-repl' to change the ClojureScript REPL to use."]
         :help "Connects to a REPL that's already running."]
        ["Start a Clojure REPL, and a ClojureScript REPL" cider-jack-in-clojurescript
         :help "Starts an nREPL server, connects a Clojure REPL to it, and then a ClojureScript REPL.
-  Configure `cider-cljs-lein-repl' to change the ClojureScript REPL to use."]
+  Configure `cider-cljs-*-repl' to change the ClojureScript REPL to use for you build tool."]
```
Comment (bbatsov) [inline cider-mode.el]: I'd list here all the actual variables. Don't like having var references to something that doesn't exist.

File: cider.el [hunk 4]
```diff
@@ -436,7 +436,9 @@ dependencies."
     ("(do (require 'cljs.repl.node) (cemerick.piggieback/cljs-repl (cljs.repl.node/repl-env)))"
      "Node" " (requires NodeJS to be installed)")
     ("(do (require 'weasel.repl.websocket) (cemerick.piggieback/cljs-repl (weasel.repl.websocket/repl-env :ip \"127.0.0.1\" :port 9001)))"
-     "Weasel" " (see Readme for additional configuration)")))
+     "Weasel" " (see Readme for additional configuration)")
```
Comment (bbatsov) [inline cider.el]: This should point to the proper manual section.

File: CHANGELOG.md (hunk 1 above)
Comment (arichiardi) [inline CHANGELOG.md]: Ok will do 😀

File: cider-mode.el (hunk 2 above)
Comment (arichiardi) [inline cider-mode.el]: Kk

File: cider.el [hunk 5]
```diff
@@ -448,6 +450,35 @@ it should start a ClojureScript REPL."
   :safe (lambda (x) (assoc x cider--cljs-repl-types))
   :group 'cider)
 
+(defcustom cider-cljs-boot-repl "(do (require 'adzerk.boot-cljs-repl) (adzerk.boot-cljs-repl/start-repl))"
```
Comment (bbatsov) [inline cider.el]: You forgot the `:package-version`. Guess the only you copied didn't have it, so you can check when it was introduced and backfill it for it as well.

**Comment by bbatsov** [comment]
Apart from the my minor remarks the PR looks great!

File: cider.el (hunk 4 above)
Comment (arichiardi) [inline cider.el]: So should I put a link to the weasel Readme? Or a link to Cider's "up and running'?

File: cider-mode.el (hunk 3 above)
Comment (arichiardi) [inline cider-mode.el]: I have used this syntax auto in other parts of the docs, should I make it explicit everywhere?

**Comment by arichiardi** [comment]
Yep and I use it now actively, it would be great to test a couple of lein (and Gradle?) projects as well

File: cider.el (hunk 4 above)
Comment (bbatsov) [inline cider.el]: Cider's "up and running".

File: cider-mode.el (hunk 3 above)
Comment (bbatsov) [inline cider-mode.el]: I'd make it explicit everywhere.

File: cider.el [hunk 6]
```diff
@@ -446,8 +448,40 @@ it should start a ClojureScript REPL."
...
+This is only used in boot projects.  It is evaluated in a Clojure REPL and
+it should start a ClojureScript REPL."
+  :type `(choice ,@(seq-map (lambda (x) `(const :tag ,(apply #'concat (cdr x)) ,(car x)))
+                            cider--cljs-repl-types)
+                 (string :tag "Custom"))
+  :safe (lambda (x) (assoc x cider--cljs-repl-types))
+  :package-version '(cider . "0.15.0")
+  :group 'cider)
+
+(defcustom cider-cljs-gradle-repl "(cemerick.piggieback/cljs-repl (cljs.repl.rhino/repl-env))"
```
Comment (expez) [inline cider.el]: What does this have to do with `gradle`?  Isn't this just how you get a cljs repl on a recent JVM?

Comment (bbatsov) [inline cider.el]: Well, that's why before there was just one defcustom regardless of the project type, but on `boot` there are other options and we decided to be flexible and give people the ability to pick different forms on a per-build tool basis.

Comment (arichiardi) [inline cider.el]: If I understand correctly, this is basically the form that cider launches on connection with a repl so in theory `gradle` users could have a task launching a `cljs` repl...very unlikely but I saw gradle everywhere and though I should add it too here :smile:

Comment (bbatsov) [inline cider.el]: Frankly from time to time I wonder why we have some special gradle support. It was requested a while back, but I doubt many people use this (if any).

Comment (arichiardi) [inline cider.el]: Probably removing `gradle` will need to be in another PR right? Do I need to do anything else for this one?

Comment (bbatsov) [inline cider.el]: Yeah, we can think about this down the road. 

Probably we can remove the hardcoded project types and add some project registration types similar to to that employed by Projectile - an API to register project types and registering some common project types by default.

**Comment by bbatsov** [comment]
👍
